tests/api_tests.cpp

```diff
@@ -0,0 +1,142 @@
+               login->calc = calc_api;
+               wsc->register_api(fc::api<login_api>(login));
+               c->set_session_data( wsc );
+          });
+
+      server->listen( 0 );
+      auto listen_port = server->get_listening_port();
+      server->start_accept();
+
+      try {
+         auto client = std::make_shared<fc::http::websocket_client>();
+         auto con  = client->connect( "ws://localhost:" + std::to_string(listen_port) );
+         server->stop_listening();
+         auto apic = std::make_shared<websocket_api_connection>(*con, MAX_DEPTH);
+         auto remote_login_api = apic->get_remote_api<login_api>();
+         auto remote_calc = remote_login_api->get_calc();
+         bool remote_triggered = false;
+         remote_calc->on_result( [&remote_triggered]( uint32_t r ) { remote_triggered = true; } );
+         BOOST_CHECK_EQUAL(remote_calc->add( 4, 5 ), 9);
+         BOOST_CHECK(remote_triggered);
+
+         client->synchronous_close();
+         server->close();
+         fc::usleep(fc::milliseconds(50));
+         client.reset();
+         server.reset();
+      } FC_LOG_AND_RETHROW()
+   } FC_LOG_AND_RETHROW()
+}
+
+BOOST_AUTO_TEST_CASE(optionals_test) {
+   try {
+      auto optionals = std::make_shared<optionals_api>();
+      fc::api<optionals_api> oapi(optionals);
+      BOOST_CHECK_EQUAL(oapi->foo("a"), "[\"a\",null,null]");
+      BOOST_CHECK_EQUAL(oapi->foo("a", "b"), "[\"a\",\"b\",null]");
+      BOOST_CHECK_EQUAL(oapi->foo("a", "b", "c"), "[\"a\",\"b\",\"c\"]");
+      BOOST_CHECK_EQUAL(oapi->foo("a", {}, "c"), "[\"a\",null,\"c\"]");
+
+      auto server = std::make_shared<fc::http::websocket_server>();
+      server->on_connection([&]( const websocket_connection_ptr& c ){
+               auto wsc = std::make_shared<websocket_api_connection>(*c, MAX_DEPTH);
+               wsc->register_api(fc::api<optionals_api>(optionals));
+               c->set_session_data( wsc );
+          });
+
+      server->listen( 0 );
+      auto listen_port = server->get_listening_port();
+      server->start_accept();
+
+      try {
+         auto client = std::make_shared<fc::http::websocket_client>();
+         auto con  = client->connect( "ws://localhost:" + std::to_string(listen_port) );
+         server->stop_listening();
+         auto apic = std::make_shared<websocket_api_connection>(*con, MAX_DEPTH);
+         auto remote_optionals = apic->get_remote_api<optionals_api>();
+
+         BOOST_CHECK_EQUAL(remote_optionals->foo("a"), "[\"a\",null,null]");
+         BOOST_CHECK_EQUAL(remote_optionals->foo("a", "b"), "[\"a\",\"b\",null]");
+         BOOST_CHECK_EQUAL(remote_optionals->foo("a", "b", "c"), "[\"a\",\"b\",\"c\"]");
+         BOOST_CHECK_EQUAL(remote_optionals->foo("a", {}, "c"), "[\"a\",null,\"c\"]");
+
+         client->synchronous_close();
+         server->close();
+         fc::usleep(fc::milliseconds(50));
+         client.reset();
+         server.reset();
+      } FC_LOG_AND_RETHROW()
```

Comment:
Ehh, in Dan's original code, it was probably there to end the scopes at the right places... or something. That code looked like a hurricane went through it.

Since I cleaned it up, they don't have any purpose, yeah. I'll remove them.